batch/batch/worker/worker.py

```diff
@@ -514,38 +473,38 @@ async def pull_image(self):
 
         try:
             if not is_google_image:
-                await self.ensure_image_is_pulled()
-            elif is_public_image:
-                auth = await self.batch_worker_access_token()
-                await self.ensure_image_is_pulled(auth=auth)
+                # Explicitly default to dockerhub
+                if self.image_ref.domain is None:
+                    self.image_ref.domain = 'docker.io'
+                    self.image_ref_str = str(self.image_ref)
+                await check_exec_output('podman', 'pull', self.image_ref_str)
             else:
-                # Pull to verify this user has access to this
-                # image.
-                # FIXME improve the performance of this with a
-                # per-user image cache.
-                auth = self.current_user_access_token()
-                await docker_call_retry(MAX_DOCKER_IMAGE_PULL_SECS, f'{self}')(
-                    docker.images.pull, self.image_ref_str, auth=auth
-                )
-        except DockerError as e:
-            if e.status == 404 and 'pull access denied' in e.message:
+                if is_public_image:
+                    auth = await self.batch_worker_access_token()
+                else:
+                    # Pull to verify this user has access to this
+                    # image.
+                    # FIXME improve the performance of this with a
+                    # per-user image cache.
+                    auth = self.current_user_access_token()
+                os.makedirs(self.container_scratch, exist_ok=True)
+                authfile_name = f'{self.container_scratch}/podman_auth.json'
+                with open(authfile_name, 'w+') as authfile:
+                    creds = base64.b64encode(f'{auth["username"]}:{auth["password"]}'.encode())
+                    auth_json = {'auths': {'gcr.io': {'auth': creds.decode()}}}
+                    authfile.write(json.dumps(auth_json))
+                await check_exec_output('podman', 'pull', f'--authfile={authfile_name}', self.image_ref_str)
```

Comment:
What does podman pull do if the image already exists? Does it still do an auth check?

Reply:
I tested this manually and it looks like on every `podman pull`, Podman checks the remote manifest to know if it needs to pull anything new, and does the authentication then. I pulled the batch worker image, ran `podman logout gcr.io/hail-vdc`, checked to ensure that the image was listed in `podman images`, and then tried to pull it again but received an unauthorized error.